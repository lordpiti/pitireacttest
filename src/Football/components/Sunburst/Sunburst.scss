.CDTtooltip {
    background-color: yellow;
    padding: 20px;
    position: absolute;
    text-align: center;
    font: 12px sans-serif;
    border: 0px;
    border-radius: 8px;
    pointer-events: none;
  }