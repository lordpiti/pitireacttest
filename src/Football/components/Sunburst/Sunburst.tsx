import { arc, hierarchy, partition } from 'd3';
import React from 'react';
import { nestData, TreeNodeType } from '../../utilities/nest';
import './Sunburst.scss';

export interface SunburstProps {
  data: TreeNodeType[];
  diameter: number;
  updateToolTip: Function;
}

interface SunburstState {
  data: any;
  rootRadius: number;
  arcHeight: number;
}

const colourSet = [
  '#a1aee3',
  '#9cd9d5',
  '#d2e4aa',
  '#f2af96',
  '#ffe080',
  '#e79ac6',
  '#908aaf',
  '#e2c6e5',
  '#bbe4ec',
  '#c7cacc',
];

export class Sunburst extends React.Component<SunburstProps, SunburstState> {
  constructor(props: SunburstProps) {
    super(props);
    this.state = {
      data: null,
      rootRadius: 0,
      arcHeight: 0,
    };
  }

  arc = arc()
    .startAngle((d: any) => d.x0)
    .endAngle((d: any) => d.x1)
    .innerRadius((d: any) =>
      d.depth === 0 ? 0 : d.depth * this.state.arcHeight
    )
    .outerRadius((d: any) => {
      let oRad: number;

      if (d.children === null && d.depth !== this.state.data.height) {
        oRad = this.state.arcHeight * (this.state.data.height + 1);
      } else {
        oRad =
          d.depth === 0
            ? this.state.rootRadius
            : this.state.arcHeight * (d.depth + 1);
      }

      return oRad;
    });
  componentDidMount() {
    this.createSunburst(this.props);
  }

  componentDidUpdate(prevProps: SunburstProps) {
    if (
      prevProps.data !== this.props.data ||
      prevProps.diameter !== this.props.diameter
    ) {
      this.createSunburst(this.props);
    }
  }

  createSunburst(props: SunburstProps) {
    const validNodes = props.data.filter(
      (x) =>
        !x.needStateId || (x.needStateId && x.products && x.products.length > 0)
    );

    const nestedData = nestData(validNodes, (treeNode: TreeNodeType) => {
      if (treeNode.products) {
        treeNode.value = treeNode.products.length;
      }
    });

    const partitionF = (data: TreeNodeType) => {
      const root = hierarchy(data).sum((d: TreeNodeType) => {
        return d.products && d.products.length;
      });
      // .sort((a: any, b: any) => b.value - a.value);
      return partition().size([2 * Math.PI, root.height + 1])(root);
    };
    const root = partitionF(nestedData);

    root.each((d: any) => (d.current = d));

    const adjustedDiameterForSunburst =
      root.height * (props.diameter / (root.height + 1));

    this.setState({
      data: root,
      arcHeight: adjustedDiameterForSunburst / (2 * (root.height + 1)),
      rootRadius: adjustedDiameterForSunburst / root.height,
    });
  }

  rtnFill = (d: any, i: number) => {
    d.index = i;
    if (!d.depth) {
      return '#ccc';
    }
    while (d.depth > 1) {
      // eslint-disable-next-line no-param-reassign
      d = d.parent;
    }

    return colourSet[d.index % colourSet.length];
  };

  render() {
    const { updateToolTip } = this.props;

    return this.state.data ? (
      <>
        {this.state.data
          .descendants()
          .slice(1)
          .map((element: any, i: number) => (
            <path
              key={`needStateArc-${element.data.id}`}
              id={`axisPath${element.data.id}`}
              fill={this.rtnFill(element, i)}
              d={this.arc(element.current) as string | undefined}
              strokeWidth={0.5}
              stroke='grey'
              onMouseMove={(event) =>
                updateToolTip(element.current.data.name, event)
              }
              onMouseLeave={(event) => updateToolTip(null, event)}
            />
          ))}
      </>
    ) : (
      <></>
    );
  }
}
