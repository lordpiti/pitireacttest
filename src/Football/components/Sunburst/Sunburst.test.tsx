import { shallow } from 'enzyme';
import React from 'react';
import { Sunburst } from './Sunburst';

import dataIn from './Sunburst.data.json';

describe('Sunburst', () => {
    const testProps = {
        data: dataIn,
        diameter: 1300,
        updateToolTip: (e, d) => undefined
    };
    const sunburst = shallow(
        <Sunburst {...testProps} />
    );

    const instance = sunburst.instance() as Sunburst;

    it('should render correctly', () => {
        expect(sunburst).toMatchSnapshot();
    });

    describe('createSunburst', () => {

        it('updates the state properly', () => {

            instance.createSunburst(testProps);

            const expectedRootHeight = 2;
            const expectedRootRadius = testProps.diameter / (expectedRootHeight + 1);
            const expectedArcHeight = (expectedRootHeight * expectedRootRadius) / (2 * (expectedRootHeight + 1));

            expect(instance.state.data.height).toEqual(expectedRootHeight);
            expect(instance.state.rootRadius).toEqual(expectedRootRadius);
            expect(instance.state.arcHeight).toEqual(expectedArcHeight);
        });
    })
});
