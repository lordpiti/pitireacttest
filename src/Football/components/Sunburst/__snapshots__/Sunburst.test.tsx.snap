// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sunburst should render correctly 1`] = `
<Fragment>
  <path
    d="M1.7689342654350655e-14,-288.88888888888886A288.88888888888886,288.88888888888886,0,0,1,169.80462844004776,233.716020597207L84.90231422002388,116.8580102986035A144.44444444444443,144.44444444444443,0,0,0,8.844671327175327e-15,-144.44444444444443Z"
    fill="#a1aee3"
    id="axisPath0"
    key="needStateArc-0"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M169.80462844004776,233.716020597207A288.88888888888886,288.88888888888886,0,0,1,-169.80462844004774,233.716020597207L-84.90231422002387,116.8580102986035A144.44444444444443,144.44444444444443,0,0,0,84.90231422002388,116.8580102986035Z"
    fill="#9cd9d5"
    id="axisPath1"
    key="needStateArc-1"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M-169.80462844004774,233.716020597207A288.88888888888886,288.88888888888886,0,0,1,-185.69419835388916,-221.30172801214914L-92.84709917694458,-110.65086400607457A144.44444444444443,144.44444444444443,0,0,0,-84.90231422002387,116.8580102986035Z"
    fill="#d2e4aa"
    id="axisPath2"
    key="needStateArc-2"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M-185.69419835388916,-221.30172801214914A288.88888888888886,288.88888888888886,0,0,1,-5.306802796305196e-14,-288.88888888888886L-2.653401398152598e-14,-144.44444444444443A144.44444444444443,144.44444444444443,0,0,0,-92.84709917694458,-110.65086400607457Z"
    fill="#f2af96"
    id="axisPath3"
    key="needStateArc-3"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M2.653401398152598e-14,-433.33333333333326A433.33333333333326,433.33333333333326,0,0,1,389.4774200629723,-189.96083027526683L259.65161337531487,-126.64055351684456A288.88888888888886,288.88888888888886,0,0,0,1.7689342654350655e-14,-288.88888888888886Z"
    fill="#a1aee3"
    id="axisPath10999"
    key="needStateArc-10999"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M389.4774200629723,-189.96083027526683A433.33333333333326,433.33333333333326,0,0,1,426.7500263052901,-75.24754365566982L284.50001753686007,-50.16502910377988A288.88888888888886,288.88888888888886,0,0,0,259.65161337531487,-126.64055351684456Z"
    fill="#a1aee3"
    id="axisPath11003"
    key="needStateArc-11003"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M426.7500263052901,-75.24754365566982A433.33333333333326,433.33333333333326,0,0,1,341.4713265629128,266.7866393077852L227.64755104194185,177.85775953852348A288.88888888888886,288.88888888888886,0,0,0,284.50001753686007,-50.16502910377988Z"
    fill="#a1aee3"
    id="axisPath11004"
    key="needStateArc-11004"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M341.4713265629128,266.7866393077852A433.33333333333326,433.33333333333326,0,0,1,254.70694266007166,350.5740308958105L169.80462844004776,233.716020597207A288.88888888888886,288.88888888888886,0,0,0,227.64755104194185,177.85775953852348Z"
    fill="#a1aee3"
    id="axisPath11009"
    key="needStateArc-11009"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M254.70694266007166,350.5740308958105A433.33333333333326,433.33333333333326,0,0,1,-90.09506602102913,423.86396031798233L-60.063377347352755,282.5759735453216A288.88888888888886,288.88888888888886,0,0,0,169.80462844004776,233.716020597207Z"
    fill="#9cd9d5"
    id="axisPath10902"
    key="needStateArc-10902"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M-90.09506602102913,423.86396031798233A433.33333333333326,433.33333333333326,0,0,1,-254.70694266007175,350.5740308958104L-169.80462844004785,233.71602059720695A288.88888888888886,288.88888888888886,0,0,0,-60.063377347352755,282.5759735453216Z"
    fill="#9cd9d5"
    id="axisPath10907"
    key="needStateArc-10907"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M-254.7069426600716,350.5740308958105A433.33333333333326,433.33333333333326,0,0,1,-401.77967031227445,162.32952381356193L-267.8531135415163,108.21968254237463A288.88888888888886,288.88888888888886,0,0,0,-169.80462844004774,233.716020597207Z"
    fill="#d2e4aa"
    id="axisPath10801"
    key="needStateArc-10801"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M-401.77967031227445,162.32952381356193A433.33333333333326,433.33333333333326,0,0,1,-389.47742006297227,-189.96083027526686L-259.65161337531487,-126.64055351684459A288.88888888888886,288.88888888888886,0,0,0,-267.8531135415163,108.21968254237463Z"
    fill="#d2e4aa"
    id="axisPath10806"
    key="needStateArc-10806"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M-389.47742006297227,-189.96083027526686A433.33333333333326,433.33333333333326,0,0,1,-278.54129753083373,-331.9525920182237L-185.69419835388916,-221.30172801214914A288.88888888888886,288.88888888888886,0,0,0,-259.65161337531487,-126.64055351684459Z"
    fill="#d2e4aa"
    id="axisPath10808"
    key="needStateArc-10808"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M-278.54129753083373,-331.9525920182237A433.33333333333326,433.33333333333326,0,0,1,-176.25254533284667,-395.8696983117937L-117.50169688856445,-263.9131322078625A288.88888888888886,288.88888888888886,0,0,0,-185.69419835388916,-221.30172801214914Z"
    fill="#f2af96"
    id="axisPath10700"
    key="needStateArc-10700"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
  <path
    d="M-176.25254533284667,-395.8696983117937A433.33333333333326,433.33333333333326,0,0,1,-7.960204194457793e-14,-433.33333333333326L-5.306802796305196e-14,-288.88888888888886A288.88888888888886,288.88888888888886,0,0,0,-117.50169688856445,-263.9131322078625Z"
    fill="#f2af96"
    id="axisPath10705"
    key="needStateArc-10705"
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    stroke="grey"
    stroke-width={0.5}
  />
</Fragment>
`;
